src/js/heuristicblocking.js

```diff
@@ -499,38 +444,91 @@ function hasCookieTracking(details, origin) {
   return false;
 }
 
+function ifRelevant(details, callback) {
+  // ignore requests that are outside a tabbed window
+  if (details.tabId < 0 || incognito.tabIsIncognito(details.tabId)) {
+    return {};
+  }
+
+  let fqdn = utils.makeURI(details.url).host,
+    origin = window.getBaseDomain(fqdn);
+
+  // if this is a main window request
+  if (details.type == "main_frame") {
+    // save the origin associated with the tab
+    log("Origin: " + origin + "\tURL: " + details.url);
+    tabOrigins[details.tabId] = origin;
+    return {};
+  }
+
+  let tabOrigin = tabOrigins[details.tabId];
+  // ignore first-party requests
+  if (!tabOrigin || origin == tabOrigin) {
+    return {};
+  }
+  return callback(details, fqdn, origin, tabOrigin);
+}
+
+function onRelevantBeforeSendHeaders(details, fqdn, origin, tabOrigin) {
+  badger.checkForDNTPolicy(fqdn, badger.storage.getNextUpdateForDomain(fqdn));
+  return heuristicBlockingAccounting(details, fqdn, origin, tabOrigin);
```

Comment:
I broke this out into a separate function because I was speculating about performance. I should measure this before it gets committed.